File: frameworks/messengers/Messenger.java
package com.amay077.stopwatchapp.frameworks.messengers;

import java.util.HashMap;
import java.util.Map;

import rx.Subscription;
import rx.functions.Action0;
import rx.functions.Action1;

public class Messenger implements Subscription {
    private final Map<String, Action1<Message>> _registerMap = new HashMap<>();

    public void send(Message message) {
        final String className = message.getClass().getName();
        if (_registerMap.containsKey(className)) {
            final Action1 handler = _registerMap.get(className);
            handler.call(message);
        }
    }

    public void register(String messageClassName, Action1<Message> action) {
        _registerMap.put(messageClassName, action);
    }

    @Override
    public void unsubscribe() {
        _registerMap.clear();
    }

    @Override
    public boolean isUnsubscribed() {
        return _registerMap.size() == 0;
    }
}

File: StopWatchAppXamarin/StopWatchApp.Core/Frameworks/Messengers/Messenger.cs
using System;
using System.Collections.Generic;

namespace StopWatchApp.Core.Frameworks.Messengers
{
	public class Messenger : IDisposable
	{
		private readonly IDictionary<string, Action<IMessage>> _registerMap = new Dictionary<string, Action<IMessage>>();

		public void Send(IMessage message)
		{
			var className = message.GetType().Name;
			if (_registerMap.ContainsKey(className)) {
				var handler = _registerMap[className];
				handler(message);
			}
		}

		public void Register(string messageClassName, Action<IMessage> action)
		{
			_registerMap.Add(messageClassName, action);
		}

		#region IDisposable implementation

		public void Dispose()
		{
			_registerMap.Clear();
		}

		#endregion
	}

}

